File: 백준/Silver/1331. 나이트 투어/나이트 투어.py
import sys

input = sys.stdin.read
data = input().strip().split("\n")

col = {
    'A': 0,
    'B': 1,
    'C': 2,
    'D': 3,
    'E': 4,
    'F': 5,
}

row = {
    '1': 0,
    '2': 1,
    '3': 2,
    '4': 3,
    '5': 4,
    '6': 5,
}

y, x = data[0][0], data[0][1]
start_row = row[x]
start_col = col[y]
prev_row = start_row
prev_col = start_col

target = set(data)
visited = [[False] * 6 for _ in range(6)]
visited[start_row][start_col] = True
answer = "Invalid"

if len(target) != 36:
    print(answer)
    sys.exit()

for i in range(1, 36):
    c, r = data[i][0], data[i][1]
    cur_x, cur_y = row[r], col[c]
    if ((abs(cur_x - prev_row) == 1 and abs(cur_y - prev_col) == 2) or
        (abs(cur_x - prev_row) == 2 and abs(cur_y - prev_col) == 1)):
        if visited[cur_x][cur_y]:
            answer = "Invalid"
            break
        visited[cur_x][cur_y] = True
        prev_row, prev_col = cur_x, cur_y
    else:
        answer = "Invalid"
        break

last_y, last_x = data[35][0], data[35][1]
if (visited[row[last_x]][col[last_y]] and
    ((abs(row[last_x] - start_row) == 1 and abs(col[last_y] - start_col) == 2) or
     (abs(row[last_x] - start_row) == 2 and abs(col[last_y] - start_col) == 1))):
    answer = "Valid"
else:
    answer = "Invalid"

print(answer)

File: 백준/Silver/1331. 나이트 투어/나이트 투어.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const col = {
  A: 0,
  B: 1,
  C: 2,
  D: 3,
  E: 4,
  F: 5,
}

const row = {
  1: 0,
  2: 1,
  3: 2,
  4: 3,
  5: 4,
  6: 5,
}

const [y, x] = input[0].split("")
let startRow = row[x]
let startCol = col[y]
let prevRow = row[x]
let prevCol = col[y]
const target = new Set(input)
const visited = Array.from({ length: 6 }, () =>
  Array.from({ length: 6 }, () => false),
)
visited[row[x]][row[y]] = true
let answer = "Invalid"

if (target.size != 36) {
  console.log(answer)
  return
}

for (let i = 1; i < 36; i += 1) {
  const [c, r] = input[i].split("")
  const [curX, curY] = [row[r], col[c]]
  if (
    (Math.abs(curX - prevRow) === 1 && Math.abs(curY - prevCol) === 2) ||
    (Math.abs(curX - prevRow) === 2 && Math.abs(curY - prevCol) === 1)
  ) {
    if (visited[curX][curY]) {
      answer = "Invalid"
      break
    }
    visited[curX][curY] = true
    prevRow = curX
    prevCol = curY
  } else {
    answer = "Invalid"
    break
  }
}
const [lastY, lastX] = input[35].split("")
if (
  visited[row[lastX]][col[lastY]] &&
  ((Math.abs(row[lastX] - startRow) === 1 &&
    Math.abs(col[lastY] - startCol) === 2) ||
    (Math.abs(row[lastX] - startRow) === 2 &&
      Math.abs(col[lastY] - startCol) === 1))
) {
  answer = "Valid"
  
} else {

  answer = "Invalid"
}
console.log(answer)
